### ISSUE TITLE
Providing fan and pump curves for the older firmware X62s

### ISSUE THREAD

**Issue by slavfox** [jonasmalacofilho/liquidctl#8]
Hi,

I have an older Kraken X62 (`iSerial 5D8A2247363`, latest firmware 2.1.8) which [allegedly doesn't and probably never will support fan control curves](https://www.reddit.com/r/NZXT/comments/9v0l11/update_nzxt_kraken_x62_firmware/e9a7l2l/). If I can help out in any way, please let me know.

I was also wondering if it'd make sense to include an option, in liquidctl, to simulate a fan control curve for older Krakens by repeatedly reading the temperature and setting the fan speed accordingly, [akin to the way a lot of existing scripts for Nvidia temperature curves work](https://gist.github.com/byaka/6337021). 

Right now, trying to set temperature crashes on https://github.com/jonasmalacofilho/liquidctl/blob/master/liquidctl/driver/kraken_two.py#L194 with a `TypeError: 'NotImplementedType' object is not callable` (should probably be `raise NotImplementedError`).

Cheers!

**Comment by jonasmalacofilho** [comment]
Hi,

While setting profiles to run straight off the device really isn't possible on the older firmware, we can certainly implement a software-based alternative.

I made a proof of concept and added it to [`extra/krakencurve-poc`](https://github.com/jonasmalacofilho/liquidctl/blob/master/extra/krakencurve-poc).

I would very much welcome patches or pull requests that take this – or rewrite it – into a proper, more polished, feature.

```
# krakencurve-poc --help
Adjust Kraken speeds dynamically, with software.

This is just a proof of concept.

Usage:
  krakencurve-poc [options] show-sensors
  krakencurve-poc [options] control <pump-profile> <fan-profile>
  krakencurve-poc --help
  krakencurve-poc --version

Options:
  --use-psutil            Enable extra sensors from psutil (requires psutil)
  --pump-sensor <sensor>  Select custom sensor for pump speed
  --fan-sensor <sensor>   Select custom sensor for fan speed
  --interval <seconds>    Update interval in seconds [default: 2]
  -n, --dry-run           Do not apply any settings
  -v, --verbose           Output additional information
  -g, --debug             Show debug information on stderr
  --version               Display the version number
  --help                  Show this message

Examples:
  krakencurve-poc control '(30,50),(40,100)' '(30,60),(45,100)'
  krakencurve-poc control '(30,50),(40,100)' '(20,25),(60,100)' --use-psutil --fan-sensor 'coretemp:Package id 0'
```

For now I picked the first Kraken X cooler found and used the liquid temperature plus (optionally) sensor data from Python's psutil, which as far as I know simply reads from the kernel hwmon files (i.e. lm-sensors data).

Besides supporting more than one device (and device type), I think the code lacks more sensor sources.  I would like to support Windows too – at least for those that, like me, dual boot and want to use the same tool on both systems – and reading temperatures from graphics cards could be very useful as well.

There is also a lack of error handling: all commands might timeout or fail with a busy error and we might need to recover/retry, or recognize if it's a persistent issue that requires more attention.

![screenshot from 2018-12-08 06-12-33](https://user-images.githubusercontent.com/1832496/49683710-48f4f000-fab0-11e8-8392-f78e3a3b305f.png)

**Comment by slavfox** [comment]
Thanks, it was really helpful to get me to figure out how you'd prefer to go about this.

I've spent today toying around with adding it to `liquidctl `under a `--software` flag, so that setting a software cooling profile is done the same way as a firmware-supported one:

```
Speed control options:
  -S --software                       Force software-controlled cooling
                                      profile, even if firmware cooling
                                      profiles are supported. Recommended
                                      only if your device doesn't support
                                      firmware cooling profiles.
  -p --use-psutil                     Enable extra sensors from psutil (requires
                                      psutil)
  -s <sensor>, --sensor <sensor>      Select custom sensor
  -i <seconds>, --interval <seconds>  Update interval in seconds [default: 2]
```

Ideally, if the daemon is running, calling eg. `liquidctl -S set fan speed  20 30  30 50  34 80  40 90  50 100` again would just update the cooling curve and trigger the daemon to reload the config; otherwise, it'd start the daemon.

Right now I'm inclined to leave the actual daemon launch script in `krakencurve-poc-control` (alternately as a `-d, --daemon` flag), and have the user themselves handle running it at system start (eg. with `systemd` or whatever their preferred solution is). The simplest cross-platform way to get it to work without adding external dependencies would probably be having the daemon listen on a socket (read from a config file?) instead of waiting for signals (I don't have a Windows install at hand, and I'd rather not write code that *may or may not work* on Windows platforms). 

This would require liquidctl to be able to read settings from a config file (which would need either adding [appdirs](https://github.com/ActiveState/appdirs) as a requirement or just borrowing the relevant parts from there), which makes me think I might as well also add a `liquidctl load config` command and support for reading all settings (colors, pump and fan speeds) from a config file while I'm at it..

I'll try to get a pull request together this week. Let me know if you have any opinions on this.

**Comment by slavfox** [comment]
> instead of -S, --software, maybe just --force-software for when it's really necessary; the other times we automatically pick the appropriate method based on if what the user wants to do is supported by the firmware or not (including the specific sensor selected)

I figured that if the user expects their hardware to handle the cooling profile, it might not be the best idea to quietly set a software one - but it actually makes perfect sense, since they'd have to manually start the daemon first. In that case, if the daemon isn't running and the hardware doesn't support software cooling profiles, we'll just show a "Your device doesn't support cooling profiles. To use a software cooling profile, start the daemon with `liquidctl start service` and rerun your command." error message and exit. 

> the same goes for --use-psutil [...] (not sure if psutil is safe enough or not yet)

Psutil reads from `hwmon` (https://github.com/giampaolo/psutil/blob/8f99f3782663959062ee868bbfdbc336307a3a4d/psutil/_pslinux.py#L1182), which has some [ problems](https://github.com/nicolargo/glances/issues/1079), and I don't think it's particularly reliable, hence why I opted to keep it all behind a flag.

> -i, --interval <seconds> is fine, but I think it should be restricted to the daemon command; 

Agreed completely, I forgot about that.

**Comment by jonasmalacofilho** [comment]
@slavfox 

How is you work on this coming along?  Is there anything you need from me?

**Comment by slavfox** [comment]
The little work I managed to do in December is on the [software-fan-curves branch in my fork](https://github.com/slavfox/liquidctl/commit/416475d155b234dad98361a9952015e31b87b8cc). The next step on my end will be to write the actual daemon controlling the temperature, and the commands for pushing config changes to it.

**Comment by jonasmalacofilho** [comment]
@slavfox, NZXT released new firmware upgrades for all third gen. Krakens (including the X62):

 - v5.1 for legacy/Rev. 1 devices
 - v6.1 for current/Rev. 2 devices

Interestingly, there hasn't been a new CAM release to accompany these.  Since it probably used a simple inequality on the firmware version to decide which features were available, I think there's a good chance v5.1 might also support profiles now.

You might want to give it a try: [Kraken X Firmware Update (02.27.19)](https://sta3-nzxtcorporation.netdna-ssl.com/uploads/download/attachment/753/KrakenXFWUpdate__02.27.19_.zip).<sup>(note 1)</sup>

And if you do, please report back with your findings.

---

Note 1: download link taken from the downloads section of the product page for the X62.

**Comment by jonasmalacofilho** [comment]
@slavfox did you manage to test the new firmware?  Does it add support for profiles?

**Comment by slavfox** [comment]
@jonasmalacofilho I found an old Windows drive, updated the firmware, got CAM running. It did in fact jump from 2.1 to 5.01, thanks for the notification.
![CAM](https://user-images.githubusercontent.com/35173529/55414931-5b69cc80-556c-11e9-882b-981cb84c925a.png)

I have tested the fan curves by setting a ridiculous curve (`liquidctl set fan speed 30 20 34 100`) and can confirm that as soon as the liquid temperature hits 34C the fans start screaming. In an extremely unexpected twist of events, thanks to NZXT themselves rev. 1 Krakens now have hardware fan curve support that works out-of-the-box with the current `liquidctl` master branch!
